File: leetcode_notes/leetcode/0001.两数之和/0001-两数之和.java
class Solution {
    // Not same element twice
    // Use HashMap to cache nums as key and index as value
    public int[] twoSum(int[] nums, int target) {
        if(nums == null || nums.length  < 2) {
            return new int[]{-1, -1};
        }
        
        int[] res = new int[]{-1, -1};
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            if(map.containsKey(target - nums[i])) {
                res[0] = map.get(target - nums[i]);
                res[1] = i;
                break;
            }
            map.put(nums[i], i);
        }
        return res;
    }
}

File: leetcode_notes/leetcode_cn/0001.两数之和/0001-两数之和.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum0(nums, target)
    nums.each_with_index do |e, i|
        idx = nums.index(target - e)
        return [i, idx] if idx && idx != i
    end
end

def two_sum(nums, target)
    map = {}
    nums.each_with_index do |e, i|
        return [map[e], i] if map.has_key?(e)
        map[target - e] = i
    end
end